Tests/LibSQL/TestSQLStorage.cpp

```diff
@@ -100,6 +101,194 @@ TEST_CASE(create_heap) {
+constexpr static u32 pointers[] = { 92, 4, 50, 47, 68, 73, 24, 28, 50, 93, 60, 36, 92, 72, 53, 26, 91, 84, 25, 43, 88, 12, 62, 35, 96, 27, 96, 27, 99, 30, 21, 89, 54, 60, 37, 68, 35, 55, 80, 2, 33, 26, 93, 70, 45, 44, 3, 66, 75, 4, };
+
+NonnullRefPtr<SQL::BTree> setup_btree(SQL::Heap& heap)
+{
+    auto index_def = SQL::IndexDef::construct("test", true, 0);
+    index_def->append_column("key_value", SQL::SQLType::Integer, SQL::SortOrder::Ascending);
+
+    auto root_pointer = heap.user_value(0);
+    if (!root_pointer) {
+        root_pointer = heap.new_record_pointer();
+        heap.set_user_value(0, root_pointer);
+    }
+    auto btree = SQL::BTree::construct(heap, index_def, root_pointer);
+    btree->on_new_root = [&]() {
+        heap.set_user_value(0, btree->root());
+    };
+    return btree;
+}
+
+void insert_and_get_to_and_from_btree(int num_keys)
+{
+    unlink("test.db");
+    {
+        auto heap = SQL::Heap::construct("test.db");
+        auto btree = setup_btree(heap);
+
+        for (auto ix = 0; ix < num_keys; ix++) {
+            SQL::Key k(btree->index_def());
+            k[0] = keys[ix];
+            k.pointer(pointers[ix]);
+            btree->insert(k);
+        }
+        btree->list_tree();
+    }
+
+    {
+        auto heap = SQL::Heap::construct("test.db");
+        auto btree = setup_btree(heap);
+
+        for (auto ix = 0; ix < num_keys; ix++) {
+            SQL::Key k(btree->index_def());
+            k[0] = keys[ix];
+            auto pointer_opt = btree->get(k);
+            EXPECT(pointer_opt.has_value());
+            EXPECT_EQ(pointer_opt.value(), pointers[ix]);
+        }
+    }
+}
+
+void insert_into_and_scan_btree(int num_keys)
+{
+    unlink("test.db");
+    {
+        auto heap = SQL::Heap::construct("test.db");
+        auto btree = setup_btree(heap);
+
+        for (auto ix = 0; ix < num_keys; ix++) {
+            SQL::Key k(btree->index_def());
+            k[0] = keys[ix];
+            k.pointer(pointers[ix]);
+            btree->insert(k);
+        }
+        btree->list_tree();
+    }
+
+    {
+        auto heap = SQL::Heap::construct("test.db");
+        auto btree = setup_btree(heap);
+
+        int count = 0;
+        SQL::Key prev;
+        for (auto iter = btree->begin(); !iter.is_end(); iter++, count++) {
+            auto key = (*iter);
+            dbgln((String) key);
+            if (prev.length()) {
+                EXPECT(prev < key);
+            }
+            auto key_value = (int) key[0];
+            for (auto ix = 0; ix < num_keys; ix++) {
+                if (keys[ix] == key_value) {
+                    EXPECT_EQ(key.pointer(), pointers[ix]);
+                    break;
+                }
+            }
+            prev = key;
+        }
+        EXPECT_EQ(count, num_keys);
+    }
+}
+
+TEST_CASE(btree_one_key)
```

Comment:
why do you make 10 tests doing essentially the same?

Reply:
Because it is my experience that with datastructures like this it's impossible to design your test so you find all corner cases. Doing it this way is a bit of a blunt force instrument but I've had good experiences with it, including in this code.